gc/base/standard/ConcurrentGCIncrementalUpdate.hpp

```diff
@@ -43,34 +89,89 @@ class MM_ConcurrentGCIncrementalUpdate : public MM_ConcurrentGC
+	uintptr_t _bytesToTracePass1;
+	uintptr_t _bytesToTracePass2;
+	uintptr_t _allocToTraceRateCardCleanPass2Boost;
+	uintptr_t _totalTracedAtPass2KO;
+	uintptr_t _totalCleanedAtPass2KO;
+	uintptr_t _traceTargetPass1;
+	uintptr_t _traceTargetPass2;
+
+	bool _pass2Started;
+	bool  _secondCardCleanPass;
 
 	/**
 	 * Creates Concurrent Card Table
 	 * @param env current thread environment
 	 * @return true if table is created
 	 */
-	bool createCardTable(MM_EnvironmentBase *env);
-	static void hookCardCleanPass2Start(J9HookInterface** hook, uintptr_t eventNum, void* eventData, void* userData);
-	void recordCardCleanPass2Start(MM_EnvironmentBase *env);
-
 public:
 	
 	/*
 	 * Function members
 	 */
+private:
+	bool createCardTable(MM_EnvironmentBase *env);
+	static void hookCardCleanPass2Start(J9HookInterface** hook, uintptr_t eventNum, void* eventData, void* userData);
+	void recordCardCleanPass2Start(MM_EnvironmentBase *env);
 
 protected:
 	bool initialize(MM_EnvironmentBase *env);
 	void tearDown(MM_EnvironmentBase *env);
+
+	void reportConcurrentFinalCardCleaningStart(MM_EnvironmentBase *env);
+	void reportConcurrentFinalCardCleaningEnd(MM_EnvironmentBase *env, uint64_t duration);
+	virtual void reportConcurrentCollectionStart(MM_EnvironmentBase *env);
+	virtual void reportConcurrentHalted(MM_EnvironmentBase *env);
+
+	void kickoffCardCleaning(MM_EnvironmentBase *env, ConcurrentCardCleaningReason reason);
+	virtual void finalConcurrentPrecollect(MM_EnvironmentBase *env);
+	virtual void tuneToHeap(MM_EnvironmentBase *env);
+	virtual void updateTuningStatisticsInternal(MM_EnvironmentBase *env);
+	virtual void signalThreadsToActivateWriteBarrierInternal(MM_EnvironmentBase *env);
+	virtual void adjustTraceTarget();
+	virtual void resetConcurrentParameters(MM_EnvironmentBase *env);
+
+	virtual uintptr_t doConcurrentTrace(MM_EnvironmentBase *env, MM_AllocateDescription *allocDescription, uintptr_t sizeToTrace, MM_MemorySubSpace *subspace, bool tlhAllocation);
+	virtual uintptr_t localMark(MM_EnvironmentBase *env, uintptr_t sizeToTrace);
+	virtual	bool cleanCards(MM_EnvironmentBase *env, bool isMutator, uintptr_t sizeToDo, uintptr_t  *sizeDone, bool threadAtSafePoint);
+
+	virtual uintptr_t getTraceTarget() {
+		if (_pass2Started) {
+			return (_traceTargetPass1 + _traceTargetPass2);
+		} else {
+			return _traceTargetPass1;
+		}
+	}
 public:
 	virtual uintptr_t getVMStateID() { return OMRVMSTATE_GC_COLLECTOR_CONCURRENTGC; };
 	virtual bool heapAddRange(MM_EnvironmentBase *env, MM_MemorySubSpace *subspace, uintptr_t size, void *lowAddress, void *highAddress);
+	virtual void J9ConcurrentWriteBarrierStoreHandler(MM_EnvironmentBase *env, omrobjectptr_t destinationObject, omrobjectptr_t storedObject);
+	virtual void J9ConcurrentWriteBarrierBatchStoreHandler(MM_EnvironmentBase *env, omrobjectptr_t destinationObject);
+#if defined(OMR_GC_MODRON_SCAVENGER)
+	virtual void oldToOldReferenceCreated(MM_EnvironmentBase *env, omrobjectptr_t objectPtr);

```

Comment:
While this all makes sense and cleans up the code, a similar comment could apply as before (I'm a bit worried about per implications of virtual dispatch). But this ok with this for now. I believe this just does not occur frequently enough. If it does, will will see it in CPU profiles (it won't be inlined)